pandas/tests/tseries/offsets/test_behavior.py

```diff
@@ -0,0 +1,334 @@
+month_strategy = st.integers(min_value=1, max_value=12)
+weekday_strategy = st.integers(min_value=0, max_value=6)
+
+
+def gen_dst_crossing():
+    # Generate either a pair of Timestamps or a date_range that is known
+    # to cross a DST transition
+    raise NotImplementedError
+
+
+def gen_date_range_freq():
+    # return a freq str or offset object suitable for passing as
+    # `freq` kwarg to date_range
+    return st.sampled_from(['Y', 'Q', 'M', 'D', 'H',
+                            'T', 's', 'ms', 'us', 'ns'])
+    # TODO: Add the rest; business, multiples, ...
+
+
+@st.composite
+def gen_random_date_range(draw):
+    # TODO: Choose the min/max values more systematically
+    start = st.datetimes(min_value=pd.Timestamp(1900, 1, 1).to_pydatetime(),
+                         max_value=pd.Timestamp(2100, 1, 1).to_pydatetime())
+    periods = st.integers(min_value=10, max_value=100)
+    freq = gen_date_range_freq()
+    tz = gen_random_tz()
+
+    dti = pd.date_range(start=draw(start), tz=draw(tz),
+                        freq=draw(freq), periods=draw(periods))
+    return dti
+
+
+def gen_random_tz():
+    # Allows None
+    return st.one_of(st.none(), hepytz.timezones())
+    # TODO: Weighting between naive and timezones?
+    # TODO: Get datetuil timezones?
+
+
+gen_random_datetime = st.datetimes(min_value=pd.Timestamp.min.to_pydatetime(),
+                                   max_value=pd.Timestamp.max.to_pydatetime(),
+                                   timezones=gen_random_tz())
+
+
+def gen_random_timestamp():
+    nano = st.integers(min_value=0, max_value=999)
+    dt = st.datetimes(min_value=pd.Timestamp.min.to_pydatetime(),
+                      max_value=pd.Timestamp.max.to_pydatetime(),
+                      timezones=gen_random_tz())
+    ts = pd.Timestamp(dt)
+
+    if dt != dt_max:
+        ts.replace(nanosecond=nano)
+    else:
+        ts = ts.replace(nanosecond=min(nano, pd.Timestamp.max.nanosecond))
+
+    # TODO: worry about timezones near min/max?
+    return ts
+
+
+def gen_random_datelike():
+    # py_dates = st.dates()
+    py_datetimes = gen_random_datetime
+
+    # dt64_dtypes = hen.datetime64_dtypes()
+    # np_dates = hen.arrays(dtype=dt64_dtypes, shape=())
+    # TODO: Allow for non-scalar versions?
+    # FIXME: dt64.__add__(offset) does not get dispatched to
+    # offset.__radd__(dt64), just raises TypeError
+
+    any_dates = st.one_of(py_datetimes)
+    return any_dates
+
+
+def gen_timedeltalike():
+    py_timedeltas = st.timedeltas(min_value=td_min, max_value=td_max)
+    pd_timedeltas = py_timedeltas.map(pd.Timedelta)
+    # TODO: get those last few nanoseconds?
+
+    td64_dtypes = hen.timedelta64_dtypes()
+    np_timedeltas = hen.arrays(dtype=td64_dtypes, shape=())
+    # TODO: Allow for non-scalar versions?
+
+    # TODO: Week
+    # TODO: Tick
+    any_tds = st.one_of(py_timedeltas, pd_timedeltas, np_timedeltas)
+    return any_tds
+
+
+@st.composite
+def gen_random_relativedelta_DateOffset(draw):
```

Comment:
I think you need to add some assumptions here to make sure that your `year`, `month`, `week`, `day`, etc values are valid. You can do this by drawing them from integers with min/max set, drawing them as you are and then adding `assume` statements, or (and this might be a bit out there), draw a random `datetime()` and just use the values from that (won't work for `week`, but you can calculate the week from the `datetime` if you want to use it, I imagine).

Reply:
Oh most definitely.  I just threw those in and figured I'd put in the appropriate bounds one-by-one as errors came up.